### PR TITLE
[Core] Run same tests for RxJava and Coroutine runtimes.

### PR DESCRIPTION
**Pull request by Laimiux**
## What
Instead of writing tests twice, I'm setting it up so same tests run for both RxJava and Coroutine runtimes.

### How
There were few changes needed to support this:
- I've decoupled `TestFormulaObserver` from RxJava runtime by introducing `FormulaTestDelegate`. For now, the RxJava will still be a default within `formula-test`.
- I've introduced `TestableRuntime` interface with `RxTestableRuntime` and `CoroutineTestableRuntime` implementations to allow switching runtimes within tests.
- I've used `@RunWith(Parameterized::class)` and `@Parameterized.Parameters` to setup `RuntimeTest` to run for both runtimes
```kotlin
@JvmStatic
@Parameterized.Parameters(name = "{1}")
fun runtimes() = listOf(
    arrayOf(RxJavaTestableRuntime, "RxJava"),
    arrayOf(CoroutinesTestableRuntime, "Coroutines"),
)
```
- I've updated all tests within `RuntimeTest` to use `TestableRuntime`.

### CODE REVIEW

**Comment by carrotkite** [comment]
### JaCoCo Code Coverage 76.9% :white_check_mark:
| Class | Covered | Meta | Status |
|:---|:---:|:---:|:---:|
| `com/instacart/formula/coroutines/FlowRuntime` | 100% | 0% | :white_check_mark: |
| `com/instacart/formula/test/RxJavaTestDelegate` | 100% | 0% | :white_check_mark: |
| `com/instacart/formula/test/TestFormulaObserver` | 50% | 0% | :white_check_mark: |


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: formula/src/test/java/com/instacart/formula/ManyEmissionStreamTest.kt [hunk 1]
```diff
@@ -6,6 +6,7 @@ import com.instacart.formula.test.test
 import io.reactivex.rxjava3.core.Observable
 import org.junit.Test
 
+// TODO:
```
Comment (Laimiux) [inline formula/src/test/java/com/instacart/formula/ManyEmissionStreamTest.kt:9]: This is a `TODO` to inspect tests that are outside of `RuntimeTest` and move them there.

File: formula/src/test/java/com/instacart/formula/ObservableFormulaTest.kt [hunk 2]
```diff
@@ -1,57 +0,0 @@
-package com.instacart.formula
-
-import com.google.common.truth.Truth
-import com.instacart.formula.rxjava3.ObservableFormula
-import com.instacart.formula.test.test
-import com.jakewharton.rxrelay3.PublishRelay
-import io.reactivex.rxjava3.core.Observable
-import org.junit.Test
-
-class ObservableFormulaTest {
```
Comment (Laimiux) [inline formula/src/test/java/com/instacart/formula/ObservableFormulaTest.kt:10]: Moved this into `RuntimeTest`

File: formula/src/test/java/com/instacart/formula/StartStopFormula.kt [hunk 3]
```diff
@@ -2,10 +2,11 @@ package com.instacart.formula
 
 import com.instacart.formula.StartStopFormula.Output
 import com.instacart.formula.StartStopFormula.State
+import com.instacart.formula.test.TestableRuntime
 
-class StartStopFormula : Formula<Unit, State, Output> {
+class StartStopFormula(runtime: TestableRuntime) : Formula<Unit, State, Output> {
 
-    val incrementEvents = IncrementRelay()
+    val incrementEvents = runtime.newIncrementRelay()
```
Comment (Laimiux) [inline formula/src/test/java/com/instacart/formula/StartStopFormula.kt:9]: Delegates to `TestableRuntime` to create `IncrementRelay` which is implemented using `FlowStream` or `RxStream` depending on the runtime.

File: formula/src/test/java/com/instacart/formula/test/TestableRuntime.kt [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+    fun streamFormula(): StreamFormulaSubject
+}
+
+interface IncrementRelay {
+    fun stream(): Stream<Unit>
+    fun triggerIncrement()
+}
+
+interface StreamFormulaSubject : IFormula<String, Int> {
```
Comment (Laimiux) [inline formula/src/test/java/com/instacart/formula/test/TestableRuntime.kt:40]: Could use a better name.

File: formula/src/test/java/com/instacart/formula/test/CoroutineTestableRuntime.kt [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    override val formula: FormulaT,
+): FormulaTestDelegate<Input, Output, FormulaT> {
+    private val values = mutableListOf<Output>()
+    private val errors = mutableListOf<Throwable>()
+
+    private val inputFlow = MutableSharedFlow<Input>(1)
+    private val formulaFlow = formula.toFlow(inputFlow)
+        .onEach { values.add(it) }
+        .onCompletion {
+            println("complete")
```
Comment (alexanderbezverhni) [inline formula/src/test/java/com/instacart/formula/test/CoroutineTestableRuntime.kt:84]: Can be removed?

File: formula/src/test/java/com/instacart/formula/test/CoroutineTestableRuntime.kt [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+import kotlinx.coroutines.test.resetMain
+import kotlinx.coroutines.test.runBlockingTest
+import kotlinx.coroutines.test.setMain
+import org.junit.rules.TestRule
+import org.junit.rules.TestWatcher
+import org.junit.runner.Description
+import org.junit.runners.model.Statement
+
+@OptIn(ExperimentalStdlibApi::class)
+object CoroutinesTestableRuntime : TestableRuntime {
```
Comment (alexanderbezverhni) [inline formula/src/test/java/com/instacart/formula/test/CoroutineTestableRuntime.kt:30]: I understand right that this runtime is still not connected and we only run 1 pass with `RxJavaTestableRuntime`?

**Review by alexanderbezverhni — APPROVED** [review]
I'm not a Coroutine/Flow expert, but this abstraction looks very reasonable to me, as compared to duplicating the whole test suite.

File: formula/src/test/java/com/instacart/formula/test/CoroutineTestableRuntime.kt (hunk 6 above)
Comment (Laimiux) [reply formula/src/test/java/com/instacart/formula/test/CoroutineTestableRuntime.kt:30]: It's already used in `FormulaRuntimeTest`

```kotlin
@RunWith(Parameterized::class)
class FormulaRuntimeTest(val runtime: TestableRuntime, val name: String) {
    companion object {
        @JvmStatic
        @Parameterized.Parameters(name = "{1}")
        fun runtimes() = listOf(
            arrayOf(RxJavaTestableRuntime, "RxJava"),
            arrayOf(CoroutinesTestableRuntime, "Coroutines"),
        )
    }

   ...
}
```

**Pull request merged**